File: 7M4W/ee/Z.py
import sys
sys.stdin = open('1074.txt')

def z(n, r, c):
    if n == 0:
        return 0
    
    half = 2 ** (n - 1)
    size = half * half

    if r < half and c < half:
        return z(n - 1, r, c)
    elif r < half and c >= half:
        return size + z(n - 1, r, c - half)
    elif r >= half and c < half:
        return 2 * size + z(n - 1, r - half, c)
    else:
        return 3 * size + z(n - 1, r - half, c - half)
    
N, R, C = map(int, input().split())
print(z(N, R, C))

File: 7M4W/ee/Z.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./1074.txt";

const input = fs.readFileSync(filePath).toString().trim().split(" ");

const [N, R, C] = input.map(Number);

function Z(n, r, c) {
  if (n === 0) {
    return 0;
  }

  const half = 2 ** (n - 1)
  const size = half * half

  if (r < half && c < half) {
    return Z(n - 1, r, c)
  } else if (r < half && c >= half) {
    return size + Z(n - 1, r, c - half)
  } else if (r >= half && c < half) {
    return 2 * size + Z(n - 1, r - half, c)
  } else {
    return 3 * size + Z(n - 1, r - half, c - half)
  }
}

console.log(Z(N, R, C))